File: 015/d.rb
require 'pp'

# W = ARGF.gets.to_i
# N, K = ARGF.gets.split(" ").map(&:to_i)
# ai = []
# bi = []
# # pi = []
# for _ in 0...N
#   a, b = ARGF.gets.split(" ").map(&:to_i)
#   ai << a
#   bi << b
#   # pi << b.to_f/a
# end
#
# # order = pi.each_with_index.sort.reverse.map(&:last)
#
# dp = Array.new(N+1) do Array.new(W+1) do Array.new(K+1, 0) end end
#
# for i in 0...N
#   for w in 1..W
#     for k in 1..K
#       if w >= ai[i] 
#         # pp dp[i][w-ai[i]][k-1]
#         dp[i+1][w][k] = [dp[i][w-ai[i]][k-1]+bi[i], dp[i][w][k]].max
#       else
#         dp[i+1][w][k] = dp[i][w][k]
#       end
#     end
#   end
# end
#
# puts dp[N][W][K]

W = ARGF.gets.to_i
N, K = ARGF.gets.split(" ").map(&:to_i)
A = []
B = []
# pi = []
for _ in 0...N
  a, b = ARGF.gets.split(" ").map(&:to_i)
  A << a
  B << b
  # pi << b.to_f/a
end

DP = Array.new(W+1) do Array.new(K+1) do Array.new(N+1) end end

def dfs(restW, restK, curN)
  return 0 if restW <= 0 || restK <= 0 || curN == N

  m = DP[restW][restK][curN]
  return m unless m.nil?

  if restW >= A[curN] && restK > 0
    res = [ B[curN] + dfs(restW-A[curN], restK-1, curN+1), dfs(restW, restK, curN+1) ].max
  else
    res = dfs(restW, restK, curN+1)
  end

  DP[restW][restK][curN] = res
  return res
end

puts dfs(W, K, 0)

File: 015/d.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	f := bufio.NewScanner(os.Stdin)
	f.Split(bufio.ScanWords)

	W := read(f)
	N := read(f)
	K := read(f)
	A := make([]int, N)
	B := make([]int, N)
	for i := range A {
		A[i] = read(f)
		B[i] = read(f)
	}

	DP := make([][][]int, W+1)
	for i := range DP {
		DP[i] = make([][]int, K+1)
		for j := range DP[i] {
			DP[i][j] = make([]int, N+1)
			for k := range DP[i][j] {
				DP[i][j][k] = -1
			}
		}
	}

	var dfs func(int, int, int) int
	dfs = func(restW int, restK int, curN int) int {
		if restW <= 0 || restK <= 0 || curN == N {
			return 0
		}

		if memo := DP[restW][restK][curN]; memo >= 0 {
			return memo
		}

		var res int
		if restW >= A[curN] && restK > 0 {
			res = max(B[curN]+dfs(restW-A[curN], restK-1, curN+1), dfs(restW, restK, curN+1))
		} else {
			res = dfs(restW, restK, curN+1)
		}

		DP[restW][restK][curN] = res
		return res
	}

	fmt.Println(dfs(W, K, 0))
}

func read(s *bufio.Scanner) int {
	s.Scan()
	res, _ := strconv.ParseInt(s.Text(), 10, 64)
	return int(res)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
